flink-connectors/flink-connector-pubsub/src/main/java/org/apache/flink/streaming/connectors/pubsub/PubSubSource.java

```diff
@@ -0,0 +1,301 @@
+		implements ResultTypeQueryable<OUT>, ParallelSourceFunction<OUT>, StoppableFunction {
+	private static final Logger LOG = LoggerFactory.getLogger(PubSubSource.class);
+	protected DeserializationSchema<OUT> deserializationSchema;
+	protected SubscriberWrapper subscriberWrapper;
+
+	protected boolean running = true;
+	protected transient volatile SourceContext<OUT> sourceContext = null;
+
+	PubSubSource(DeserializationSchema<OUT> deserializationSchema, SubscriberWrapper subscriberWrapper) {
+		super(String.class);
+		this.deserializationSchema = deserializationSchema;
+		this.subscriberWrapper = subscriberWrapper;
+	}
+
+	@Override
+	public void open(Configuration configuration) throws Exception {
+		super.open(configuration);
+		subscriberWrapper.initialize();
+		if (hasNoCheckpointingEnabled(getRuntimeContext())) {
+			throw new IllegalArgumentException("The PubSubSource REQUIRES Checkpointing to be enabled and " +
+				"the checkpointing frequency must be MUCH lower than the PubSub timeout for it to retry a message.");
+		}
+
+		getRuntimeContext().getMetricGroup().gauge("PubSubMessagesProcessedNotAcked", this::getOutstandingMessagesToAck);
+		getRuntimeContext().getMetricGroup().gauge("PubSubMessagesReceivedNotProcessed", subscriberWrapper::amountOfMessagesInBuffer);
+	}
+
+	private boolean hasNoCheckpointingEnabled(RuntimeContext runtimeContext) {
+		return !(runtimeContext instanceof StreamingRuntimeContext && ((StreamingRuntimeContext) runtimeContext).isCheckpointingEnabled());
+	}
+
+	@Override
+	protected void acknowledgeSessionIDs(List<AckReplyConsumer> ackReplyConsumers) {
+		ackReplyConsumers.forEach(AckReplyConsumer::ack);
+	}
+
+	@Override
+	public void run(SourceContext<OUT> sourceContext) throws Exception {
+		this.sourceContext = sourceContext;
+		subscriberWrapper.start();
+
+		while (subscriberWrapper.isRunning()) {
+			try {
+				Tuple2<PubsubMessage, AckReplyConsumer> newMessage = subscriberWrapper.take();
+				if (newMessage != null) {
+					processMessage(newMessage);
+				}
+			} catch (InterruptedException e) {
+				LOG.debug("Interrupted - stop or cancel called?");
+			}
+		}
+
+		nackOutstandingMessages();
+
+		LOG.debug("Waiting for PubSubSubscriber to terminate.");
+		subscriberWrapper.awaitTerminated();
+	}
+
+	void processMessage(Tuple2<PubsubMessage, AckReplyConsumer> input) {
+		PubsubMessage message = input.f0;
+		AckReplyConsumer ackReplyConsumer = input.f1;
+		if (sourceContext == null) {
+			ackReplyConsumer.nack();
+			return;
+		}
+
+		synchronized (sourceContext.getCheckpointLock()) {
+			boolean alreadyProcessed = !addId(message.getMessageId());
+			if (alreadyProcessed) {
+				return;
+			}
+
+			sessionIds.add(ackReplyConsumer);
+			OUT deserializedMessage = deserializeMessage(message);
+			sourceContext.collect(deserializedMessage);
+			if (deserializationSchema.isEndOfStream(deserializedMessage)) {
```

Comment:
This is not in line with the definition of `isEndOfStream()`: `Method to decide whether the element signals the end of the stream. If true is returned the element won't be emitted.`.
The element is collected in any case in the previous statement.

Reply:
Fixed